prow/cmd/gangway/main.go

```diff
@@ -0,0 +1,225 @@
+	if c.dryRun {
+		return job, nil
+	}
+	return c.client.Create(ctx, job, o)
+}
+
+// interruptableServer is a wrapper type around the gRPC server, so that we can
+// pass it along to our own interrupts package.
+type interruptableServer struct {
+	grpcServer *grpc.Server
+	listener   net.Listener
+	port       int
+}
+
+// Shutdown shuts down the inner gRPC server as gracefully as possible, by first
+// invoking GracefulStop() on it. This gives the server time to try to handle
+// things gracefully internally. However if it takes too long (if the parent
+// context cancels us), we forcefully kill the server by calling Stop(). Stop()
+// interrupts GracefulStop() (see
+// https://pkg.go.dev/google.golang.org/grpc#Server.Stop).
+func (s *interruptableServer) Shutdown(ctx context.Context) error {
+
+	gracefulStopFinished := make(chan struct{})
+
+	go func() {
+		s.grpcServer.GracefulStop()
+		close(gracefulStopFinished)
+	}()
+
+	select {
+	case <-gracefulStopFinished:
+		return nil
+	case <-ctx.Done():
+		s.grpcServer.Stop()
+		return ctx.Err()
+	}
+}
+
+func (s *interruptableServer) ListenAndServe() error {
+	logrus.Infof("serving gRPC on port %d", s.port)
+	return s.grpcServer.Serve(s.listener)
+}
+
+func main() {
+	logrusutil.ComponentInit()
+
+	o := gatherOptions(flag.NewFlagSet(os.Args[0], flag.ExitOnError), os.Args[1:]...)
+	if err := o.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid options")
+	}
+
+	configAgent, err := o.config.ConfigAgent()
+	if err != nil {
+		logrus.WithError(err).Fatal("Error starting config agent.")
+	}
+
+	prowjobClient, err := o.client.ProwJobClient(configAgent.Config().ProwJobNamespace, o.dryRun)
+	if err != nil {
+		logrus.WithError(err).Fatal("unable to create prow job client")
+	}
+	kubeClient := &kubeClient{
+		client: prowjobClient,
+		dryRun: o.dryRun,
+	}
+
+	// If we are provided credentials for Git hosts, use them. These credentials
+	// hold per-host information in them so it's safe to set them globally.
+	if o.cookiefilePath != "" {
+		cmd := exec.Command("git", "config", "--global", "http.cookiefile", o.cookiefilePath)
+		if err := cmd.Run(); err != nil {
+			logrus.WithError(err).Fatal("unable to set cookiefile")
+		}
+	}
+
+	gitClient, err := o.github.GitClientFactory(o.cookiefilePath, &o.config.InRepoConfigCacheDirBase, o.dryRun)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting Git client.")
+	}
+	cacheGetter, err := config.NewInRepoConfigCacheHandler(o.config.InRepoConfigCacheSize, configAgent, gitClient, o.config.InRepoConfigCacheCopies)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error creating InRepoConfigCacheGetter.")
+	}
+	metrics.ExposeMetrics("gangway", configAgent.Config().PushGateway, o.instrumentationOptions.MetricsPort)
+
+	// Start serving liveness endpoint /healthz.
+	health := pjutil.NewHealthOnPort(o.instrumentationOptions.HealthPort)
```

Comment:
@cjwagner This part is new vs https://github.com/kubernetes/test-infra/pull/28036 (liveness endpoint).